### PR TITLE
Added GlobalExceptionHandler and Validation

### PR DESCRIPTION
**Pull request by AlexVulpine**
Implemented GlobalExceptionHandler. Added meaningful Validation for existing database. Updated application.properties to application.yml to make it more readable.

### CODE REVIEW

**Review by Parashchuk — APPROVED** [review]
LGTM

File: pom.xml [hunk 1]
```diff
@@ -71,6 +71,12 @@
             <version>${querydsl.version}</version>
             <classifier>jakarta</classifier>
         </dependency>
+
+        <dependency>
+            <groupId>org.hibernate</groupId>
+            <artifactId>hibernate-validator</artifactId>
+            <version>${hibernate-validator.version}</version>
+        </dependency>
```
Comment (Andriy-dz) [inline pom.xml]: it is better to use spring-boot-starter-validation

File: src/main/java/com/foalex/bookstore/dto/CreateBookRequestDto.java [hunk 2]
```diff
@@ -1,19 +1,34 @@
...
+    @Length(max = 128)
     private String author;
 
+    @NotNull
+    @UniqueElements
+    @Length(max = 128)
     private String isbn;
 
+    @NotNull
+    @Positive
```
Comment (Andriy-dz) [inline src/main/java/com/foalex/bookstore/dto/CreateBookRequestDto.java]: NotNulL ZeroOrPositive

File: src/main/java/com/foalex/bookstore/exception/GlobalExceptionHandler.java [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+            HttpStatusCode status, WebRequest request
+    ) {
+        Map<String, Object> body = createResponseBody(HttpStatus.BAD_REQUEST,
+                ex.getBindingResult().getAllErrors());
+        return new ResponseEntity<>(body, headers, status);
+    }
+
+    @ExceptionHandler(value = {EntityNotFoundException.class,
+                               DataProcessingException.class,
+                               DataIntegrityViolationException.class})
```
Comment (Andriy-dz) [inline src/main/java/com/foalex/bookstore/exception/GlobalExceptionHandler.java]: let's split this logic into several methods to add specific comments

**Review by Andriy-dz — CHANGES_REQUESTED** [review]

File: pom.xml [hunk 4]
```diff
@@ -16,6 +16,8 @@
     <properties>
         <lombok.mapstruct.binding.version>0.2.0</lombok.mapstruct.binding.version>
         <mapstruct.version>1.5.5.Final</mapstruct.version>
+        <maven.compiler.source>1.8</maven.compiler.source>
+        <maven.compiler.target>1.8</maven.compiler.target>
         <maven.checkstyle.plugin.configLocation>checkstyle.xml</maven.checkstyle.plugin.configLocation>
         <java.version>17</java.version>
```
Comment (NazarSmith228) [inline pom.xml]: you're trying to use Java 8 and Java 17 simultaneously :)

File: src/main/java/com/foalex/bookstore/controller/BookController.java [hunk 5]
```diff
@@ -34,13 +35,14 @@ public BookDto getBookById(@PathVariable Long id) {
...
     @ResponseStatus(HttpStatus.CREATED)
-    public BookDto createBook(@RequestBody CreateBookRequestDto bookDto) {
+    public BookDto createBook(@RequestBody @Valid CreateBookRequestDto bookDto) {
         return bookService.save(bookDto);
     }
 
     @PutMapping("/{id}")
     @ResponseStatus(HttpStatus.ACCEPTED)
-    public BookDto updateBook(@PathVariable Long id, @RequestBody CreateBookRequestDto bookDto) {
+    public BookDto updateBook(@PathVariable Long id,
```
Comment (NazarSmith228) [inline src/main/java/com/foalex/bookstore/controller/BookController.java:47]: you can also validate `@PathVariable` like:
```
@PathVariable @Positive Long id
```

in order to do so - annotate you controller with `@Validated`

File: src/main/java/com/foalex/bookstore/dto/CreateBookRequestDto.java [hunk 6]
```diff
@@ -1,19 +1,34 @@
...
 
+import jakarta.validation.constraints.Min;
+import jakarta.validation.constraints.NotNull;
 import java.math.BigDecimal;
 import lombok.Data;
+import org.hibernate.validator.constraints.Length;
+import org.hibernate.validator.constraints.UniqueElements;
 
 @Data
 public class CreateBookRequestDto {
```
Comment (NazarSmith228) [inline src/main/java/com/foalex/bookstore/dto/CreateBookRequestDto.java]: pls use record

if necessary - update your checkstyle plugin to 3.3.0

File: src/main/java/com/foalex/bookstore/dto/CreateBookRequestDto.java [hunk 7]
```diff
@@ -1,19 +1,34 @@
...
+    @NotNull
```
Comment (NazarSmith228) [inline src/main/java/com/foalex/bookstore/dto/CreateBookRequestDto.java]: it is better to use `@NotBlank` with strings

File: src/main/java/com/foalex/bookstore/dto/CreateBookRequestDto.java [hunk 8]
```diff
@@ -1,19 +1,34 @@
...
+    @Length(max = 128)
     private String title;
 
+    @NotNull
+    @Length(max = 128)
     private String author;
 
+    @NotNull
+    @UniqueElements
```
Comment (NazarSmith228) [inline src/main/java/com/foalex/bookstore/dto/CreateBookRequestDto.java]: how does this constraint work? 

this property is unique on DB layer

File: src/main/java/com/foalex/bookstore/dto/CreateBookRequestDto.java [hunk 9]
```diff
@@ -1,19 +1,34 @@
...
+    @Length(max = 128)
     private String isbn;
 
+    @NotNull
+    @Min(0)
```
Comment (NazarSmith228) [inline src/main/java/com/foalex/bookstore/dto/CreateBookRequestDto.java]: can your book cost zero ? :)

if yes - you can use `@PositiveOrZero`

otherwise - `@Positive`

File: src/main/java/com/foalex/bookstore/exception/GlobalExceptionHandler.java [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+@ControllerAdvice
+public class GlobalExceptionHandler extends ResponseEntityExceptionHandler {
+    private static final Map<Class<? extends RuntimeException>, HttpStatus> HTTP_STATUS_MAP;
+
+    static {
+        HTTP_STATUS_MAP = new HashMap<>();
+        HTTP_STATUS_MAP.put(EntityNotFoundException.class, HttpStatus.NOT_FOUND);
+        HTTP_STATUS_MAP.put(DataProcessingException.class, HttpStatus.BAD_REQUEST);
+        HTTP_STATUS_MAP.put(DataIntegrityViolationException.class, HttpStatus.CONFLICT);
+    }
```
Comment (NazarSmith228) [inline src/main/java/com/foalex/bookstore/exception/GlobalExceptionHandler.java:30]: this is completely redundant :)

don't struggle with overengineering

File: src/main/java/com/foalex/bookstore/exception/GlobalExceptionHandler.java [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+            MethodArgumentNotValidException ex, HttpHeaders headers,
+            HttpStatusCode status, WebRequest request
+    ) {
+        Map<String, Object> body = createResponseBody(HttpStatus.BAD_REQUEST,
+                ex.getBindingResult().getAllErrors());
+        return new ResponseEntity<>(body, headers, status);
+    }
+
+    @ExceptionHandler(value = EntityNotFoundException.class)
+    protected ResponseEntity<Object> handleEntityNotFoundException(RuntimeException ex) {
```
Comment (NazarSmith228) [inline src/main/java/com/foalex/bookstore/exception/GlobalExceptionHandler.java:43]: this is a handler for `EntityNotFound`

so why the parameter is `RuntimeException`?
```suggestion
    @ExceptionHandler(value = EntityNotFoundException.class)
    protected ResponseEntity<Object> handleEntityNotFoundException(EntityNotFoundException ex) {

```

File: src/main/java/com/foalex/bookstore/exception/GlobalExceptionHandler.java [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+                                              ? getErrorMessage((ObjectError) err)
+                                              : ((RuntimeException) err).getMessage())
+                                      .toList();
+        body.put("errors", messages);
+        return body;
+    }
+
+    private HttpStatus determineHttpStatus(RuntimeException ex) {
+        return HTTP_STATUS_MAP.getOrDefault(ex.getClass(), HttpStatus.INTERNAL_SERVER_ERROR);
+    }
```
Comment (NazarSmith228) [inline src/main/java/com/foalex/bookstore/exception/GlobalExceptionHandler.java:78]: this is redundant

status should be coupled to error handler method

File: src/main/java/com/foalex/bookstore/exception/GlobalExceptionHandler.java [hunk 13]
```diff
@@ -0,0 +1,85 @@
...
+    protected ResponseEntity<Object> handleDataIntegrityViolationException(RuntimeException ex) {
+        HttpStatus status = determineHttpStatus(ex);
+        Map<String, Object> body = createResponseBody(status, List.of(ex));
+        return new ResponseEntity<>(body, status);
+    }
+
+    private Map<String, Object> createResponseBody(HttpStatus status, List<?> errors) {
+        Map<String, Object> body = new LinkedHashMap<>();
+        body.put("timestamp", LocalDateTime.now());
+        body.put("status", status);
```
Comment (NazarSmith228) [inline src/main/java/com/foalex/bookstore/exception/GlobalExceptionHandler.java:66]: it is redundant to store status in response body

File: src/main/resources/db/changelog/changes/01-create-books-table.yaml [hunk 14]
```diff
@@ -36,13 +36,13 @@ databaseChangeLog:
                     nullable: false
               - column:
                   name: description
-                  type: varchar(1024)
+                  type: varchar(1028)
```
Comment (NazarSmith228) [inline src/main/resources/db/changelog/changes/01-create-books-table.yaml:39]: is it safe to update existing Liquibase migration?

doesn't it throw exception in such case?

**Review by NazarSmith228 — CHANGES_REQUESTED** [review]
Good job! Let's improve your solution

File: src/main/java/com/foalex/bookstore/exception/GlobalExceptionHandler.java (hunk 13 above)
Comment (AlexVulpine) [reply src/main/java/com/foalex/bookstore/exception/GlobalExceptionHandler.java:66]: I have resolved it in Spring Security hw

File: src/main/java/com/foalex/bookstore/exception/GlobalExceptionHandler.java (hunk 11 above)
Comment (AlexVulpine) [reply src/main/java/com/foalex/bookstore/exception/GlobalExceptionHandler.java:43]: I have resolved it in Spring Security hw

File: src/main/java/com/foalex/bookstore/exception/GlobalExceptionHandler.java (hunk 10 above)
Comment (AlexVulpine) [reply src/main/java/com/foalex/bookstore/exception/GlobalExceptionHandler.java:30]: I have resolved it in Spring Security hw

File: src/main/java/com/foalex/bookstore/controller/BookController.java [hunk 15]
```diff
@@ -34,19 +38,20 @@ public BookDto getBookById(@PathVariable Long id) {
...
-    public BookDto createBook(@RequestBody CreateBookRequestDto bookDto) {
+    public BookDto createBook(@RequestBody @Valid CreateBookRequestDto bookDto) {
         return bookService.save(bookDto);
     }
 
     @PutMapping("/{id}")
-    @ResponseStatus(HttpStatus.ACCEPTED)
-    public BookDto updateBook(@PathVariable Long id, @RequestBody CreateBookRequestDto bookDto) {
+    @ResponseStatus(HttpStatus.OK)
+    public BookDto updateBook(@PathVariable Long id,
```
Comment (liliia-ponomarenko) [inline src/main/java/com/foalex/bookstore/controller/BookController.java:47]:
```suggestion
    public BookDto updateBook(@PathVariable @Positive Long id,

```

File: src/main/resources/db/changelog/changes/01-create-books-table.yaml (hunk 14 above)
Comment (liliia-ponomarenko) [inline src/main/resources/db/changelog/changes/01-create-books-table.yaml:39]: Why did you change it?

**Review by liliia-ponomarenko — APPROVED** [review]
I see your changes in Security PR. But please next tome make it in current PR.

File: src/main/resources/db/changelog/changes/01-create-books-table.yaml (hunk 14 above)
Comment (AlexVulpine) [reply src/main/resources/db/changelog/changes/01-create-books-table.yaml:39]: I probably made some changes, when was implementing validation. But then figured out it was wrong decision, so I reversed changes and dropped db at that time. Although it seems I reversed this one incorrectly, so now it remains with current DB.
